test-suite/curvestates.cpp

```diff
@@ -86,92 +131,206 @@ struct CommonVars {
             todaysDiscounts[i] = todaysDiscounts[i-1] /
                 (1.0+todaysForwards[i-1]*accruals[i-1]);
 
-        // Coterminal swap rates & annuities
-        Size N = todaysForwards.size();
-        todaysCoterminalSwapRates = std::vector<Rate>(N);
-        coterminalAnnuity = std::vector<Real>(N);
-        Real floatingLeg = 0.0;
-        for (Size i=1; i<=N; ++i) {
-            if (i==1) {
-                coterminalAnnuity[N-1] = accruals[N-1]*todaysDiscounts[N];
-            } else {
-                coterminalAnnuity[N-i] = coterminalAnnuity[N-i+1] +
-                    accruals[N-i]*todaysDiscounts[N-i+1];
-            }
-            floatingLeg = todaysDiscounts[N-i]-todaysDiscounts[N];
-            todaysCoterminalSwapRates[N-i] =
-                floatingLeg/coterminalAnnuity[N-i];
-        }
+        // taus & first alive rates
+        std::vector<Time> evolutionTimes(N);
+        std::copy(rateTimes.begin(), rateTimes.end() - 1, evolutionTimes.begin());
+        EvolutionDescription evolution(rateTimes, evolutionTimes);
+        taus = evolution.rateTaus();
+        firstAliveRates = evolution.firstAliveRate();
 
-        std::vector<Time> evolutionTimes(rateTimes.size()-1);
-        std::copy(rateTimes.begin(), rateTimes.end()-1,
-                  evolutionTimes.begin());
-        EvolutionDescription evolution(rateTimes,evolutionTimes);
-        evolution.rateTaus();
-        evolution.firstAliveRate();
     }
 };
 
+ BOOST_AUTO_TEST_CASE(testLMMCurveState) {
+
+     BOOST_TEST_MESSAGE("Testing Libor-market-model curve state...");
+
+     CommonVars vars;
+
+     LMMDriftCalculator lmmDriftcalculator(vars.pseudo, vars.displacements, vars.taus,
+                                           vars.numeraire, vars.firstAliveRates[0]);
+     LMMCurveState lmmCs(vars.rateTimes);
+     lmmCs.setOnForwardRates(vars.todaysForwards);
+
+     std::vector<Real> lmmDrifts(vars.N);
+     lmmDriftcalculator.compute(lmmCs, lmmDrifts);
+
+    // for (Size i = 0; i < vars.N; ++i) {
+    //    std::cout << "LMM drifts: " << lmmDrifts[i] << std::endl;
+    //    std::cout << "LMM discounts ratios: " << lmmCs.discountRatio(i, vars.N) << std::endl;
+    //    std::cout << "LMM forward rates: " << lmmCs.forwardRate(i) << std::endl;
+    // }
```

Comment:
I'd clean up these commented lines (and the others below) before merging.  They're convenient during development of the tests, but are mostly unused afterwards and risk becoming out of date.

Reply:
Deleted those comments